### PR TITLE
Make solution models

### LINKED ISSUE #5: Make the solution model
What has to be done:
The solution data model has to be created containing all the properties of the data model

Why it has to be done:
A model has to be created that will represent the solution object, containing all the different data fields so that it can be stored in a database and fetched with an API request

When it is done:
When a solution model is created and contains all the fields needed to store the solution data. This has to match the requrements of both the GGC and RTR apps.

### PR DESCRIPTION
**Pull request by EGKangaroo**
- making the models is very tedious and not very fun and if I have to test literally every property of these models I will turn off sonar way

closes #5

### CODE REVIEW

**Comment by MindHawk** [comment]
> I will turn off sonar way
this is not the wae

**Review by LarsVDriel — APPROVED** [review]

File: src/Models/Like.cs
Comment (MindHawk) [inline src/Models/Like.cs:15]: Why is this virtual?

File: src/Models/Material.cs
Comment (MindHawk) [inline src/Models/Material.cs:15]: Does this relationship have to be bidirectional? Are we ever requesting the solution from a material?

File: src/Models/ResponseModels/MaterialResponse.cs
Comment (MindHawk) [inline src/Models/ResponseModels/MaterialResponse.cs:8]: Does this actually need to be a separate ResponseModel if we remove the bidirectional relationship in the database model? Or do we want to enforce strict segregation between database and response models? If so, why?

File: src/Models/ResponseModels/SDGResponse.cs
Comment (MindHawk) [inline src/Models/ResponseModels/SDGResponse.cs:11]: Shouldn't this be an enum? Potentially a string?

File: src/Models/ResponseModels/StepResponseModel.cs
Comment (MindHawk) [inline src/Models/ResponseModels/StepResponseModel.cs:11]: Should this be a string?

File: src/Models/Tool.cs
Comment (MindHawk) [inline src/Models/Tool.cs:10]: Why does tool not have a solution, but the other sub-models do?

File: src/Models/Solution.cs
Comment (MindHawk) [inline src/Models/Solution.cs:10]: I'm not sure we need to apply inheritance here. Are different solutions significantly different enough between them, or are empty fields acceptable? Inheritance adds a lot of complexity and handling, where we could otherwise choose to never return empty fields instead.

File: src/Data/SolutionsServiceContext.cs
Comment (MindHawk) [inline src/Data/SolutionsServiceContext.cs:17]: DbSet is managed by entity framework, right?

**Review by MindHawk — CHANGES_REQUESTED** [review]

File: src/Models/Like.cs
Comment (EGKangaroo) [reply src/Models/Like.cs:15]: https://stackoverflow.com/questions/8542864/why-use-virtual-for-class-properties-in-entity-framework-model-definitions

File: src/Models/Material.cs
Comment (EGKangaroo) [reply src/Models/Material.cs:15]: For Entity Framework this is necessary for proper mapping

File: src/Models/ResponseModels/SDGResponse.cs
Comment (EGKangaroo) [reply src/Models/ResponseModels/SDGResponse.cs:11]: Enums cannot be content managed easily. The Goals could be reviewed and reworded annually

File: src/Models/ResponseModels/MaterialResponse.cs
Comment (EGKangaroo) [reply src/Models/ResponseModels/MaterialResponse.cs:8]: We should not remove the bidirectional relationship in the database model + validation on request/response level should not mix with database validation.

File: src/Models/ResponseModels/StepResponseModel.cs
Comment (EGKangaroo) [reply src/Models/ResponseModels/StepResponseModel.cs:11]: What else?

File: src/Models/ResponseModels/SDGResponse.cs
Comment (MindHawk) [reply src/Models/ResponseModels/SDGResponse.cs:11]: In this case, there's also the danger they can be renumbered

File: src/Models/Solution.cs
Comment (EGKangaroo) [reply src/Models/Solution.cs:10]: Not applying inheritance implies we need to do a LOT of validation on request to ensure that fields that need to be empty are in fact empty. This would actually lead to much more complexity and inefficiency later down the road.

File: src/Models/Tool.cs
Comment (EGKangaroo) [reply src/Models/Tool.cs:10]: Because I forgot

File: src/Models/ResponseModels/StepResponseModel.cs
Comment (MindHawk) [reply src/Models/ResponseModels/StepResponseModel.cs:11]: There's an Image class but that's probably not storable in a database. I presume this is just the link? Might want to reflect that it's a link to the image in the name e.g. `CoverImageURL`

File: src/Models/Solution.cs
Comment (MindHawk) [reply src/Models/Solution.cs:10]: I'm not sure it's more complex, since all of the validation on the fields that are allowed to be filled will always be the same and not different, making the validation on the empty fields as easy as a `if(SolutionType=x)`, which could easily be one method used on all of the fields being validated, since all it cares about in this case is if they're empty or not. Or it could be as easy as an attribute on the field (https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/concepts/attributes/)

Having to cast and check and handle inheritance on a DB level seems like it'll add more complexity than that. Though if we think in the future dozens more types of solutions will be made this'll be true.

Comment (EGKangaroo) [reply src/Models/Solution.cs:10]: On the DB level, things are handled entirely by Entity Framework which means none of the complexity has to be handled by us

File: src/Models/ResponseModels/SDGResponse.cs
Comment (EGKangaroo) [reply src/Models/ResponseModels/SDGResponse.cs:11]: It's not very easy for a user, like an admin, to add a whole new value to an enum rather than storing the list of SDGs into a database that can be added to and updated.

File: src/Models/Like.cs
Comment (MindHawk) [inline src/Models/Like.cs:7]: Microsoft docs recommends System.Text.Json(.Serialization), is there a specific reason to using the newtonsoft package? Does the default json (de)serializer not support ignoring specific fields? I'm not sure what the difference between the two is

**Review by LarsVDriel — APPROVED** [review]

File: test/SolutionsTest.csproj
Comment (MindHawk) [inline test/SolutionsTest.csproj]: happi noise

**Review by MindHawk — CHANGES_REQUESTED** [review]
Should review the model with the other group before this is merged

File: src/Models/Solution.cs
Comment (MindHawk) [inline src/Models/Solution.cs:12]: C# classes do not inherit constructors. If this constructor is unused by the child classes it should be deleted. Else it should be protected instead of public.

File: src/Models/Like.cs
Comment (EGKangaroo) [reply src/Models/Like.cs:7]: https://stackoverflow.com/a/65600926

Comment (EGKangaroo) [reply src/Models/Like.cs:7]: fixed in next commit

File: src/Models/Solution.cs
Comment (EGKangaroo) [reply src/Models/Solution.cs:12]: fixed in next commit

**Review by LarsVDriel — APPROVED** [review]

**Review by MindHawk — CHANGES_REQUESTED** [review]
Remove the newtonsoft package reference and I'll approve

**Review by LarsVDriel — APPROVED** [review]

**Review by MindHawk — APPROVED** [review]
If it breaks, it's not my fault \o/

**Pull request merged**
